theano/gpuarray/pool.c

```diff
@@ -0,0 +1,341 @@
+    const DTYPE_i0* x_slice = x + offset;
+    DTYPE_o0 collector = 0;
+
+    for (ga_size d=dstart; d < dend; ++d) {
+      for (ga_size h=hstart; h < hend; ++h) {
+        for (ga_size w=wstart; w < wend; ++w) {
+          collector += x_slice[(d * height + h) * width + w];
+        }
+      }
+    }
+    if (sum_mode) {
+      z[index] = collector;
+    }
+    else {
+      z[index] = collector / pool_size;
+    }
+  }
+}
+
+#section support_code
+
+// CUDA: number of blocks for threads.
+inline int GET_BLOCKS(const int nkernels, const int nthreads) {
+  return (nkernels + nthreads - 1) / nthreads;
+}
+
+// output shape for a given input padded shape, window shape and stride
+#define OUTPUT_DIMS(in_dim, ws, st)                       \
+  (IGNORE_BORDER ? (in_dim - ws)/st + 1 :                 \
+   (st > ws ? (in_dim - 1)/st + 1 :                       \
+    std::max<size_t>(0, (in_dim - 1 - ws + st)/st) + 1))
+
+#section support_code_struct
+
+int APPLY_SPECIFIC(pool)(PyGpuArrayObject *x,
+                         PyArrayObject *ws,
+                         PyArrayObject *stride,
+                         PyArrayObject *pad,
+                         PyGpuArrayObject **z,
+                         PyGpuContextObject *ctx) {
+  if (!GpuArray_IS_C_CONTIGUOUS(&x->ga))
+    {
+      PyErr_Format(PyExc_ValueError,
+                   "GpuPool: requires data to be C-contiguous");
+      return 1;
+    }
+  size_t ndims = PyArray_DIM(ws, 0);
+  if (PyGpuArray_NDIM(x) != ndims + 2)
+    {
+      PyErr_SetString(PyExc_ValueError, "GpuPool: rank error");
+      return 1;
+    }
+  // prepare output
+  const size_t* x_dims = PyGpuArray_DIMS(x);
+  size_t z_dims[5]; // avoid warning if use 2 + nd
+  size_t w[3];
+  size_t s[3];
+  size_t p[3]; z_dims[0] = x_dims[0]; z_dims[1] = x_dims[1];
+  for (int i = 0; i < ndims; i++) {
+    w[i] = *((npy_intp*)PyArray_GETPTR1(ws, i));
+    s[i] = *((npy_intp*)PyArray_GETPTR1(stride, i));
+    p[i] = *((npy_intp*)PyArray_GETPTR1(pad, i));
+    z_dims[2 + i] = OUTPUT_DIMS(x_dims[2 + i] + 2*p[i], w[i], s[i]);
+  }
+
+  if (theano_prep_output(z, PyGpuArray_NDIM(x), z_dims,
+                         x->ga.typecode, GA_C_ORDER, ctx) != 0)
+    {
+      PyErr_SetString(PyExc_RuntimeError,
+                      "GpuPool: failed to allocate memory");
+      return 1;
+    }
+  {
+    // scope for running kernel
+    size_t max_threads_dim;
+    int err;
+
+    // get the max threads per blocks
+    err = gpucontext_property(ctx->ctx, GA_CTX_PROP_MAXLSIZE0, &max_threads_dim);
+    if (err != GA_NO_ERROR){
+      PyErr_SetString(PyExc_RuntimeError, "Could not fetch max_threads_dims");
+      return 1;
+    }
+    size_t threads_per_block = max_threads_dim;
+
+    if (ndims == 2) {
+      size_t num_kernels = z_dims[0] * z_dims[1] * z_dims[2] * z_dims[3];
+      size_t n_blocks = GET_BLOCKS(num_kernels, threads_per_block);
+      if (MAX_POOL) {
+        err = max_pool2d_kernel_call(1, &n_blocks, &threads_per_block, 0,
```

Comment:
I would use `_scall` here rather than the crude scheduling that mimics CAFFE.  I found that performance is often better when you don't try to maximise the number of threads, but rather strive for a reasonable balance with threads and blocks.

Anyway, the `_scall` version takes care of scheduling an appropriate number of blocks and threads as long as your kernel is ready to deal with unmatched number of blocks/threads, which is the case here. So I recommend that.

The interface is `_scall(1, &num_kernels, 0, <kernel parameters>`  The first parameter is the dimension of the launch (only 1d works for now).  The second parameter is the number of elements to compute in total.  The third one is the amount of shared memory to allocate.